docs/source/llms/openai/notebooks/openai-embeddings-generation.ipynb

```diff
@@ -0,0 +1,458 @@
+    "- The `insert_space_after_tags` function is called within `extract_text_from_url` to improve text readability post-extraction.\n",
+    "\n",
+    "\n",
+    "These functions are integral to preprocessing web content, ensuring that the text fed into the embedding model is clean, relevant, and well-structured."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def insert_space_after_tags(soup, tags):\n",
+    "    \"\"\"\n",
+    "    Insert a space after each tag specified in the provided BeautifulSoup object.\n",
+    "\n",
+    "    :param soup: BeautifulSoup object representing the parsed HTML.\n",
+    "    :param tags: List of tag names (as strings) after which space should be inserted.\n",
+    "    \"\"\"\n",
+    "    for tag_name in tags:\n",
+    "        for tag in soup.find_all(tag_name):\n",
+    "            tag.insert_after(\" \")\n",
+    "\n",
+    "\n",
+    "def extract_text_from_url(url, id):\n",
+    "    \"\"\"\n",
+    "    Extract and return text content from a specific section of a webpage.\n",
+    "\n",
+    "    The function targets a div with class 'section' and id 'llms', then extracts\n",
+    "    text from <h>, <li>, and <p> tags, excluding <p> tags within <ul> and any\n",
+    "    <li> tags that contain <p> with <a> having class 'reference external'. It also\n",
+    "    excludes any standalone 'Note' entries.\n",
+    "\n",
+    "    :param url: URL of the webpage from which to extract text.\n",
+    "    :param id: The target id for the div containing the main text content of the page\n",
+    "    :return: A string containing the extracted text or an error message.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    response = requests.get(url)\n",
+    "\n",
+    "    if response.status_code == 200:\n",
+    "        soup = BeautifulSoup(response.text, \"html.parser\")\n",
+    "\n",
+    "        # Find the div with class 'section' and id 'llms'\n",
+    "        target_div = soup.find(\"div\", {\"class\": \"section\", \"id\": id})\n",
+    "\n",
+    "        if target_div:\n",
+    "            # Make href and strong tags more readable\n",
+    "            insert_space_after_tags(target_div, [\"strong\", \"a\"])\n",
+    "\n",
+    "            content_tags = []\n",
+    "            for tag in target_div.find_all([\"h1\", \"h2\", \"h3\", \"h4\", \"h5\", \"h6\", \"li\", \"p\"]):\n",
+    "                if (\n",
+    "                    tag.name == \"li\"\n",
+    "                    and tag.find(\"p\")\n",
+    "                    and tag.find(\"a\", class_=\"reference external\")\n",
+    "                ):\n",
+    "                    continue  # Skip this specific <li>\n",
+    "                content_tags.append(tag)\n",
+    "\n",
+    "            return \"\\n\".join(\n",
+    "                tag.get_text(separator=\" \", strip=True)\n",
```

Comment:
optional - I feel this one-line code is a bit too long for one-line, would prefer writing it as a for-loop block.

Reply:
switched to a comprehension